app/src/main/java/ai/elimu/appstore/BaseApplication.java

```diff
@@ -99,4 +121,53 @@ public Retrofit getRetrofit() {
 
         return retrofit;
     }
+
+    /**
+     * Provides {@SecurePreferences} singleton instance to access shared preferences
+     *
+     * @return A single instance of {@SecurePreferences}
+     */
+    public static SharedPreferences getSharedPreferences() {
+        if (sSecurePreferences == null) {
+            synchronized (BaseApplication.class) {
+                if (sSecurePreferences == null) {
+                    sSecurePreferences = new SecurePreferences(getAppContext(), PREF_PASSWORD,
+                            PREF_FILE_NAME);
+                    SecurePreferences.setLoggingEnabled(BuildConfig.DEBUG);
+                }
+            }
+        }
+        return sSecurePreferences;
+    }
+
+    public static Context getAppContext() {
+        return BaseApplication.sContext;
+    }
+
+    /**
+     * Get keystore hash value to use as secure preferences' password
+     * @return The hash value generated from signing key
+     */
+    private String getKeyHash(){
+        String keyHash = "";
+        try {
+            PackageInfo info = getPackageManager().getPackageInfo(getPackageName(),
+                    PackageManager.GET_SIGNATURES);
+            for (Signature signature : info.signatures) {
+                MessageDigest md = MessageDigest.getInstance("SHA");
+                md.update(signature.toByteArray());
+                keyHash = Base64.encodeToString(md.digest(), Base64.DEFAULT);
+            }
+
+            return keyHash;
+        }
+        catch (PackageManager.NameNotFoundException e) {
+            e.printStackTrace();
```

Comment:
Do not use `e.printStackTrace();`

Reply:
Can you tell me the reason for not using it?

Reply:
Because it always logs the error, without allowing Timber to manage it. Timber helps us format all log output, and decides when to log something or not.

Reply:
Ok. I have replaced the log.